### PR TITLE
Memlet Tree Fixes and Offline SDFG Building

### PR DESCRIPTION
**Pull request by phschaad**
- [x] Fix an infinite recursion in memlet tree building
- [x] Various smaller fixes to enable offline SDFG Building.

### CODE REVIEW

**Comment by tbennun** [comment]
@phschaad could this also fix edges on the memlet tree not highlighting sometimes?

**Comment by phschaad** [comment]
@tbennun First experiments seem to indicate that it could, yes - do you remember any specific instances where it did _not_ work so I can make doubly sure?

File: src/renderer/renderer.ts [hunk 1]
```diff
@@ -52,25 +52,29 @@ declare const canvas2pdf: any;
...
-    type: string | null, foreground_elem: SDFGElement | undefined | null,
-    lib: any, _mousepos: any
+    type: SDFGElementType | null,
+    foreground_elem: SDFGElement | undefined | null, lib: any, _mousepos: any
 ): boolean {
     if (type !== null) {
         switch (type) {
-            case 'SDFGState':
+            case SDFGElementType.SDFGState:
                 return (foreground_elem instanceof NestedSDFG ||
```
Comment (tbennun) [inline src/renderer/renderer.ts:71]: uh shouldn't this be nested SDFG?

**Review by tbennun — CHANGES_REQUESTED** [review]
One comment, otherwise LGTM

File: src/renderer/renderer.ts (hunk 1 above)
Comment (phschaad) [reply src/renderer/renderer.ts:71]: Why? It's checking if the element you're trying to add it to is either a nested SDFG or the top-level SDFG (so no element)

**Review by tbennun — APPROVED** [review]

**Pull request merged**